src/main/java/org/tugraz/sysds/runtime/data/TensorBlock.java

```diff
@@ -17,23 +17,148 @@
+import org.tugraz.sysds.runtime.io.IOUtilFunctions;
+import org.tugraz.sysds.runtime.matrix.data.MatrixBlock;
 import org.tugraz.sysds.runtime.util.UtilFunctions;
 
-public abstract class TensorBlock implements CacheBlock
-{
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+
+public class TensorBlock implements CacheBlock, Externalizable {
+	private static final long serialVersionUID = -8768054067319422277L;
+
 	public static final int[] DEFAULT_DIMS = new int[]{0, 0};
+	public static final ValueType DEFAULT_VTYPE = ValueType.FP64;
+
+	private int[] _dims;
+	private boolean _heterogeneous;
+
+	private DataTensorBlock _dataTensor = null;
+	private BasicTensorBlock _basicTensor = null;
+
+	public TensorBlock() {
+		this(DEFAULT_DIMS, false);
+	}
+
+	public TensorBlock(int[] dims, boolean heterogeneous) {
+		_dims = dims;
+		_heterogeneous = heterogeneous;
+	}
+
+	public TensorBlock(int[] dims, ValueType vt) {
+		this(dims, false);
+		_basicTensor = new BasicTensorBlock(vt, dims, false);
+	}
+
+	public TensorBlock(int[] dims, ValueType[] schema) {
+		this(dims, true);
+		_dataTensor = new DataTensorBlock(schema, dims);
+	}
+
+	public TensorBlock(double value) {
+		_dims = new int[]{1, 1};
+		_heterogeneous = false;
+		_basicTensor = new BasicTensorBlock(value);
+	}
+
+	public TensorBlock(BasicTensorBlock basicTensor) {
+		this(basicTensor._dims, false);
+		_basicTensor = basicTensor;
+	}
+
+	public TensorBlock(DataTensorBlock dataTensor) {
+		this(dataTensor._dims, true);
+		_dataTensor = dataTensor;
+	}
 
-	//min 2 dimensions to preserve proper matrix semantics
-	protected int[] _dims; //[2,inf)
+	public TensorBlock(TensorBlock that) {
+		copy(that);
+	}
+
+	public void reset() {
+		if (_heterogeneous) {
+			if (_dataTensor == null)
+				_dataTensor = new DataTensorBlock(DEFAULT_VTYPE, _dims);
+			_dataTensor.reset();
+		}
+		else {
+			if (_basicTensor == null)
+				_basicTensor = new BasicTensorBlock(DEFAULT_VTYPE, _dims);
+			_basicTensor.reset();
+		}
+	}
+
+	public void reset(int[] dims) {
+		_dims = dims;
+		if (_heterogeneous) {
+			if (_dataTensor == null)
+				_dataTensor = new DataTensorBlock(DEFAULT_VTYPE, _dims);
+			_dataTensor.reset(dims);
+		}
+		else {
+			if (_basicTensor == null)
+				_basicTensor = new BasicTensorBlock(DEFAULT_VTYPE, _dims);
+			_basicTensor.reset(dims);
+		}
+	}
 
-	public abstract void reset();
+	public boolean isHeterogeneous() {
+		return _heterogeneous;
+	}
 
-	public abstract void reset(int[] dims);
+	public boolean isAllocated() {
+		if (_heterogeneous)
+			return _dataTensor != null && _dataTensor.isAllocated();
+		else
+			return _basicTensor != null && _basicTensor.isAllocated();
+	}
 
-	public abstract boolean isAllocated();
+	public TensorBlock allocateBlock() {
+		if (_heterogeneous) {
+			if (_dataTensor == null)
```

Comment:
Please make sure the individual tensorblocks are not unnecessarily reallocated.

Reply:
check for reallocation happens inside `BasicTensorBlock` and `DataTensorBlock`, although they should maybe be reworked. I think unnecessarily reallocations will not happen.